File: app/src/main/java/net/galmiza/android/spectrogram/FrequencyView.java
/**
 * Spectrogram Android application
 * Copyright (c) 2013 Guillaume Adam  http://www.galmiza.net/

 * This software is provided 'as-is', without any express or implied warranty.
 * In no event will the authors be held liable for any damages arising from the use of this software.
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it freely,
 * subject to the following restrictions:

 * 1. The origin of this software must not be misrepresented; you must not claim that you wrote the original software. If you use this software in a product, an acknowledgment in the product documentation would be appreciated but is not required.
 * 2. Altered source versions must be plainly marked as such, and must not be misrepresented as being the original software.
 * 3. This notice may not be removed or altered from any source distribution.
 */

package net.galmiza.android.spectrogram;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.util.AttributeSet;
import android.view.MotionEvent;
import android.view.View;

import androidx.annotation.Nullable;

/**
 * Class associated with the spectrogram view
 * Handles events:
 *  onSizeChanged, onTouchEvent, onDraw
 */
public class FrequencyView extends View {
    
    // Attributes
    private final Activity activity;
    private final Paint paint = new Paint();
    private @Nullable Bitmap bitmap;
    private @Nullable Canvas canvas;
    private int pos;
    private int samplingRate;
    private int width, height;
    private float[] magnitudes;
    private final int[] colorRainbow = new int[] {    0xFFFFFFFF, 0xFFFF00FF, 0xFFFF0000, 0xFFFFFF00, 0xFF00FF00, 0xFF00FFFF, 0xFF0000FF, 0xFF000000 };
    private final int[] colorFire = new int[] {    0xFFFFFFFF, 0xFFFFFF00, 0xFFFF0000, 0xFF000000 };
    private final int[] colorIce = new int[] {    0xFFFFFFFF, 0xFF00FFFF, 0xFF0000FF, 0xFF000000 };
    private final int[] colorGrey = new int[] {    0xFFFFFFFF, 0xFF000000 };

    public FrequencyView(Context context) {
        super(context);
        activity = (Activity) Misc.getAttribute("activity");
    }
    public FrequencyView(Context context, AttributeSet attrs) {
        super(context, attrs);
        activity = (Activity) Misc.getAttribute("activity");
    }
    
    
    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
        width = w;
        height = h;
        if (bitmap!=null)    bitmap.recycle();
        if (width > 0 && height > 0) {
            bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888);
            canvas = new Canvas(bitmap);
        } else {
            bitmap = null;
            canvas = null;
        }
    }
    @SuppressLint("ClickableViewAccessibility")
	@Override
    public boolean onTouchEvent(MotionEvent event) {
        invalidate();
        return true;
    }
    
    /**
     * Simple sets
     */
    public void setFFTResolution(int res) {
        magnitudes = new float[res];
    }
    public void setSamplingRate(int sampling) {
        samplingRate = sampling;
    }
    public void setMagnitudes(float[] m) {
        System.arraycopy(m, 0, magnitudes, 0, m.length);
    }
    
    /**
     * Called whenever a redraw is needed
     * Renders spectrogram and scale on the right
     * Frequency scale can be linear or logarithmic
     */
    @Override
    public void onDraw(Canvas canvas) {
        int[] colors = null;
        String colorScale = Misc.getPreference(activity, "color_scale", activity.getString(R.string.preferences_color_scale_default_value));
        switch (colorScale) {
            case "Grey":
                colors = colorGrey;
                break;
            case "Fire":
                colors = colorFire;
                break;
            case "Ice":
                colors = colorIce;
                break;
            case "Rainbow":
                colors = colorRainbow;
                break;
        }

        int wColor = 10;
        int wFrequency = 30;
        int rWidth = width-wColor-wFrequency;
        paint.setStrokeWidth(1);
        
        // Get scale preferences
		String defFrequency = activity.getString(R.string.preferences_frequency_scale_default_value);
    	boolean logFrequency = !Misc.getPreference("frequency_scale", defFrequency).equals(defFrequency);
        
        // Update buffer bitmap
        paint.setColor(Color.BLACK);
        if (this.canvas != null) {
            this.canvas.drawLine(pos%rWidth, 0, pos%rWidth, height, paint);
        }
        for (int i=0; i<height; i++) {
        	float j = getValueFromRelativePosition((float)(height-i)/height, (float)samplingRate/2, logFrequency);
        	j /= (double) samplingRate /2;
        	float mag = magnitudes[(int) (j*magnitudes.length/2)];
            float db = (float) Math.max(0,-20*Math.log10(mag));
            int c = getInterpolatedColor(colors, db*0.009f);
            paint.setColor(c);
            int x = pos%rWidth;
            this.canvas.drawPoint(x, i, paint);
            this.canvas.drawPoint(x, i, paint); // make color brighter
            //this.canvas.drawPoint(pos%rWidth, height-i, paint); // make color even brighter
        }
        
        // Draw bitmap
        if (bitmap != null) {
            if (pos<rWidth) {
                canvas.drawBitmap(bitmap, wColor, 0, paint);
            } else {
                canvas.drawBitmap(bitmap, (float) wColor - pos%rWidth, 0, paint);
                canvas.drawBitmap(bitmap, (float) wColor + (rWidth - pos%rWidth), 0, paint);
            }
        }


        // Draw color scale
        paint.setColor(Color.BLACK);
        canvas.drawRect(0, 0, wColor, height, paint);
        for (int i=0; i<height; i++) {
            int c = getInterpolatedColor(colors, ((float) i)/height);
            paint.setColor(c);
            canvas.drawLine(0, i, wColor-5, i, paint);
        }
        
        // Draw frequency scale
        float ratio = 0.7f*getResources().getDisplayMetrics().density;
        paint.setTextSize(12f*ratio);
        paint.setColor(Color.BLACK);
        canvas.drawRect(rWidth + wColor, 0, width, height, paint);
        paint.setColor(Color.WHITE);
        canvas.drawText("kHz", rWidth + wColor, 12*ratio, paint);
        if (logFrequency) {
        	for (int i=1; i<5; i++) {
	    		float y = getRelativePosition((float) Math.pow(10,i), (float)samplingRate/2);
	    		canvas.drawText("1e"+i, rWidth + wColor, (1f-y)*height, paint);
    		}
        } else {
	        for (int i=0; i<(samplingRate-500)/2; i+=1000)
	            canvas.drawText(" "+i/1000, rWidth + wColor, height*(1f-(float) i/((float)samplingRate/2)), paint);
        }
        
        pos++;
    }

    
    /**
     * Converts relative position of a value within given boundaries
     * Log=true for logarithmic scale
     */
    private float getRelativePosition(float value, float maxValue) {
        return ((float) Math.log10(1+value- (float) 1) / (float) Math.log10(1+maxValue- (float) 1));
    }
    
    /**
     * Returns a value from its relative position within given boundaries
     * Log=true for logarithmic scale
     */
    private float getValueFromRelativePosition(float position, float maxValue, boolean log) {
    	if (log)	return (float) (Math.pow(10, position*Math.log10(1+maxValue- (float) 1))+ (float) 1 -1);
    	else		return (float) 1 + position*(maxValue- (float) 1);
    }
    
    /**
     * Calculate rainbow colors
     */
    private int ave(int s, int d, float p) {
        return s + Math.round(p * (d - s));
    }
    public int getInterpolatedColor(int[] colors, float unit) {
        if (unit <= 0) return colors[0];
        if (unit >= 1) return colors[colors.length - 1];
        
        float p = unit * (colors.length - 1);
        int i = (int) p;
        p -= i;

        // now p is just the fractional part [0...1) and i is the index
        int c0 = colors[i];
        int c1 = colors[i + 1];
        int a = ave(Color.alpha(c0), Color.alpha(c1), p);
        int r = ave(Color.red(c0), Color.red(c1), p);
        int g = ave(Color.green(c0), Color.green(c1), p);
        int b = ave(Color.blue(c0), Color.blue(c1), p);

        return Color.argb(a, r, g, b);
    }
    
}

File: app/src/main/java/net/galmiza/android/spectrogram/FrequencyView.kt
/**
 * Spectrogram Android application
 * Copyright (c) 2013 Guillaume Adam  http://www.galmiza.net/
 *
 * This software is provided 'as-is', without any express or implied warranty.
 * In no event will the authors be held liable for any damages arising from the use of this software.
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it freely,
 * subject to the following restrictions:
 *
 * 1. The origin of this software must not be misrepresented; you must not claim that you wrote the original software. If you use this software in a product, an acknowledgment in the product documentation would be appreciated but is not required.
 * 2. Altered source versions must be plainly marked as such, and must not be misrepresented as being the original software.
 * 3. This notice may not be removed or altered from any source distribution.
 */
package net.galmiza.android.spectrogram

import net.galmiza.android.spectrogram.Misc.getPreference
import android.graphics.Bitmap
import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.util.AttributeSet
import android.view.MotionEvent
import android.view.View
import kotlin.math.log10
import kotlin.math.pow
import kotlin.math.roundToInt

/**
 * Class associated with the spectrogram view
 * Handles events:
 * onSizeChanged, onTouchEvent, onDraw
 */
class FrequencyView : View {
    // Attributes
    private val paint = Paint()
    private var bitmap: Bitmap? = null
    private var canvas: Canvas? = null
    private var pos = 0
    private var samplingRate = 0
    private var _width = 0
    private var _height = 0
    private lateinit var _magnitudes: FloatArray
    private val colorRainbow =
        intArrayOf(-0x1, -0xff01, -0x10000, -0x100, -0xff0100, -0xff0001, -0xffff01, -0x1000000)
    private val colorFire = intArrayOf(-0x1, -0x100, -0x10000, -0x1000000)
    private val colorIce = intArrayOf(-0x1, -0xff0001, -0xffff01, -0x1000000)
    private val colorGrey = intArrayOf(-0x1, -0x1000000)

    constructor(context: Context?) : super(context)
    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs)

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        _width = w
        _height = h
        bitmap?.recycle()
        if (width > 0 && height > 0) {
            val bitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ARGB_8888)
            this.bitmap = bitmap
            canvas = Canvas(bitmap)
        } else {
            bitmap = null
            canvas = null
        }
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        val touch = super.onTouchEvent(event)
        performClick()
        return touch
    }

    override fun performClick(): Boolean {
        val perform =  super.performClick()
        invalidate()
        return perform
    }

    /**
     * Simple sets
     */
    fun setFFTResolution(res: Int) {
        _magnitudes = FloatArray(res)
    }

    fun setSamplingRate(sampling: Int) {
        samplingRate = sampling
    }

    fun setMagnitudes(m: FloatArray) {
        System.arraycopy(m, 0, _magnitudes, 0, m.size)
    }

    /**
     * Called whenever a redraw is needed
     * Renders spectrogram and scale on the right
     * Frequency scale can be linear or logarithmic
     */
    public override fun onDraw(canvas: Canvas) {
        var colors: IntArray? = null
        val colorScale = getPreference(
            context,
            "color_scale",
            context.getString(R.string.preferences_color_scale_default_value)
        )

        colorScale?.let {
            when (it) {
                "Grey" -> colors = colorGrey
                "Fire" -> colors = colorFire
                "Ice" -> colors = colorIce
                "Rainbow" -> colors = colorRainbow
            }
        }

        val wColor = 10
        val wFrequency = 30
        val rWidth = width - wColor - wFrequency
        paint.strokeWidth = 1f

        // Get scale preferences
        val defFrequency = context.getString(R.string.preferences_frequency_scale_default_value)
        val frequency = getPreference(context, "frequency_scale", defFrequency)

        val logFrequency = frequency?.let { frequency != defFrequency } ?: false

        // Update buffer bitmap
        paint.color = Color.BLACK
        this.canvas?.drawLine(
            (pos % rWidth).toFloat(),
            0f,
            (pos % rWidth).toFloat(),
            height.toFloat(),
            paint
        )

        for (i in 0 until height) {
            var j = getValueFromRelativePosition(
                (height - i).toFloat() / height,
                samplingRate.toFloat() / 2,
                logFrequency
            )
            j /= (samplingRate.toDouble() / 2).toFloat()
            val mag = _magnitudes[(j * _magnitudes.size / 2).toInt()]
            val db = 0.0.coerceAtLeast(-20 * log10(mag.toDouble())).toFloat()
            val c = getInterpolatedColor(colors, db * 0.009f)
            paint.color = c
            val x = pos % rWidth

            this.canvas?.let {
                it.drawPoint(x.toFloat(), i.toFloat(), paint)
                it.drawPoint(x.toFloat(), i.toFloat(), paint) // make color brighter
                //this.canvas.drawPoint(pos%rWidth, height-i, paint); // make color even brighter
            }
        }

        // Draw bitmap
        bitmap?.let {
            if (pos < rWidth) {
                canvas.drawBitmap(it, wColor.toFloat(), 0f, paint)
            } else {
                canvas.drawBitmap(it, wColor.toFloat() - pos % rWidth, 0f, paint)
                canvas.drawBitmap(it, wColor.toFloat() + (rWidth - pos % rWidth), 0f, paint)
            }
        }

        // Draw color scale
        paint.color = Color.BLACK
        canvas.drawRect(0f, 0f, wColor.toFloat(), height.toFloat(), paint)
        for (i in 0 until height) {
            val c = getInterpolatedColor(colors, i.toFloat() / height)
            paint.color = c
            canvas.drawLine(0f, i.toFloat(), (wColor - 5).toFloat(), i.toFloat(), paint)
        }

        // Draw frequency scale
        val ratio = 0.7f * resources.displayMetrics.density
        paint.textSize = 12f * ratio
        paint.color = Color.BLACK
        canvas.drawRect((rWidth + wColor).toFloat(), 0f, width.toFloat(), height.toFloat(), paint)
        paint.color = Color.WHITE
        canvas.drawText("kHz", (rWidth + wColor).toFloat(), 12 * ratio, paint)
        if (logFrequency) {
            for (i in 1..4) {
                val y = getRelativePosition(
                    10.0.pow(i.toDouble()).toFloat(),
                    samplingRate.toFloat() / 2
                )
                canvas.drawText("1e$i", (rWidth + wColor).toFloat(), (1f - y) * height, paint)
            }
        } else {
            var i = 0
            while (i < (samplingRate - 500) / 2) {
                canvas.drawText(
                    " " + i / 1000,
                    (rWidth + wColor).toFloat(),
                    height * (1f - i.toFloat() / (samplingRate.toFloat() / 2)),
                    paint
                )
                i += 1000
            }
        }
        pos++
    }

    /**
     * Converts relative position of a value within given boundaries
     * Log=true for logarithmic scale
     */
    private fun getRelativePosition(value: Float, maxValue: Float): Float {
        return log10((1 + value - 1f).toDouble())
            .toFloat() / log10((1 + maxValue - 1f).toDouble()).toFloat()
    }

    /**
     * Returns a value from its relative position within given boundaries
     * Log=true for logarithmic scale
     */
    private fun getValueFromRelativePosition(
        position: Float,
        maxValue: Float,
        log: Boolean
    ): Float {
        return if (log) (10.0.pow(position * log10((1 + maxValue - 1f).toDouble())) + 1f - 1).toFloat() else 1f + position * (maxValue - 1f)
    }

    /**
     * Calculate rainbow colors
     */
    private fun ave(s: Int, d: Int, p: Float): Int {
        return s + (p * (d - s)).roundToInt()
    }

    private fun getInterpolatedColor(colors: IntArray?, unit: Float): Int {
        colors?.let {
            if (unit <= 0) return it[0]
            if (unit >= 1) return it[it.size - 1 ]
            var p = unit * (it.size - 1)
            val i = p.toInt()
            p -= i.toFloat()

            // now p is just the fractional part [0...1) and i is the index
            val c0 = it[i]
            val c1 = it[i + 1]
            val a = ave(Color.alpha(c0), Color.alpha(c1), p)
            val r = ave(Color.red(c0), Color.red(c1), p)
            val g = ave(Color.green(c0), Color.green(c1), p)
            val b = ave(Color.blue(c0), Color.blue(c1), p)
            return Color.argb(a, r, g, b)
        }

        return  0
    }
}